internal/p2p/channel.go

```diff
@@ -0,0 +1,208 @@
+// and some should ban the peer.
+//
+// FIXME: This should probably be replaced by a more general PeerBehavior
+// concept that can mark good and bad behavior and contributes to peer scoring.
+// It should possibly also allow reactors to request explicit actions, e.g.
+// disconnection or banning, in addition to doing this based on aggregates.
+type PeerError struct {
+	NodeID types.NodeID
+	Err    error
+}
+
+func (pe PeerError) Resolve() error { return fmt.Errorf("peer=%q: %w", pe.NodeID, pe.Err) }
+func (pe PeerError) Error() string  { return pe.Resolve().Error() }
+
+// Channel is a bidirectional channel to exchange Protobuf messages with peers,
+// wrapped in Envelope to specify routing info (i.e. sender/receiver).
+type Channel struct {
+	ID    ChannelID
+	In    <-chan Envelope  // inbound messages (peers to reactors)
+	Out   chan<- Envelope  // outbound messages (reactors to peers)
+	Error chan<- PeerError // peer error reporting
+
+	messageType proto.Message // the channel's message type, used for unmarshaling
+}
+
+// NewChannel creates a new channel. It is primarily for internal and test
+// use, reactors should use Router.OpenChannel().
+func NewChannel(
+	id ChannelID,
+	messageType proto.Message,
+	inCh <-chan Envelope,
+	outCh chan<- Envelope,
+	errCh chan<- PeerError,
+) *Channel {
+	return &Channel{
+		ID:          id,
+		messageType: messageType,
+		In:          inCh,
+		Out:         outCh,
+		Error:       errCh,
+	}
+}
+
+func (ch *Channel) Send(ctx context.Context, envelope Envelope) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case ch.Out <- envelope:
+		return nil
+	}
+}
+
+func (ch *Channel) SendError(ctx context.Context, pe PeerError) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case ch.Error <- pe:
+		return nil
+	}
+}
+
+func (ch *Channel) Receive(ctx context.Context) *ChannelIterator {
+	iter := &ChannelIterator{
+		pipe: make(chan Envelope), // unbuffered
+	}
+	go func() {
+		defer close(iter.pipe)
+		iteratorWorker(ctx, ch, iter.pipe)
+	}()
+	return iter
+}
+
+// ChannelIterator provides a context-aware path for callers
+// (reactors) to process messages from the P2P layer without relying
+// on the implementation details of the P2P layer. Channel provides
+// access to it's Outbound stream as an iterator, and the
+// MergedChannelIterator makes it possible to combine multiple
+// channels into a single iterator.
+type ChannelIterator struct {
+	pipe    chan Envelope
+	current *Envelope
+}
+
+func iteratorWorker(ctx context.Context, ch *Channel, pipe chan Envelope) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case envelope := <-ch.In:
+			select {
+			case <-ctx.Done():
+				return
+			case pipe <- envelope:
+			}
+		}
+	}
+}
+
+// Next returns true when the Envelope value has advanced, and false
+// when the context is canceled or iteration should stop, then Next
+// will return true. In general, use Next, as in:
```

Comment:
```suggestion
// Next returns true when the Envelope value has advanced, and false
// when the context is canceled or iteration should stop. If an iterator has returned false,
// it will never return true again.
// in general, use Next, as in:
```